File: C/216-combination-sum-iii/main.go
package main

import "fmt"

func combinationSum3(k int, n int) [][]int {
	var recur func(k, n, s int) [][]int
	recur = func(k, n, s int) [][]int {
		if k == 1 {
			if s <= n && s < 10 && n < 10 {
				return [][]int{{n}}
			}
			return nil
		}
		var ret [][]int
		for i := s; i < 10; i++ {
			if i > n {
				continue
			}
			tr := recur(k-1, n-i, i+1)
			for m := 0; m < len(tr); m++ {
				t := make([]int, len(tr[m])+1)
				copy(t[1:], tr[m])
				t[0] = i
				ret = append(ret, t)
			}
		}
		return ret
	}
	return recur(k, n, 1)
}

func main() {
	fmt.Println("a")
}

File: C/216-combination-sum-iii/main.rs
struct Solution {}

impl Solution {
  pub fn combination_sum3(k: i32, n: i32) -> Vec<Vec<i32>> {
    fn dfs(cur: i32, arr: Vec<i32>, k: i32, n: i32, ret: &mut Vec<Vec<i32>>) {
      let mut a = arr.clone();
      a.push(cur);
      if arr.iter().sum::<i32>() > n {
        return;
      }

      if a.iter().sum::<i32>() == n {
        if a.len() == k as usize {
          ret.push(a);
        }
        return;
      }

      (cur + 1..10).for_each(|d| {
        dfs(d, a.clone(), k, n, ret);
      });
    }

    let mut ret: Vec<Vec<i32>> = Vec::new();
    (1..10).for_each(|d| {
      dfs(d, Vec::new(), k, n, &mut ret);
    });
    ret
  }
}
